app/javascript/src/javascripts/post_flags.js

```diff
@@ -1,17 +1,176 @@
+  let form = null;
+  if (noteField) {
+    form = noteField.closest("form");
+  } else {
+    const dialogBody = document.querySelector(".flag-dialog-body");
+    if (dialogBody) {
+      form = dialogBody.querySelector("form");
+    }
+  }
+  let errorMessage = null;
+
+  // Parse dataset boolean values robustly. Treat presence or values like "", "true", "1", "yes" as true.
+  function parseDatasetBoolean (val) {
+    if (val === undefined || val === null) return false;
+    const v = String(val).trim().toLowerCase();
+    if (v === "" || v === "true" || v === "1" || v === "yes") return true;
+    return false;
+  }
+
+  function updateNoteRequired () {
+    try {
+      if (!noteField) return;
+      const selected = document.querySelector("input[name=\"post_flag[reason_name]\"]:checked");
+      const label = noteField.closest(".flag-notes")?.querySelector("label");
+      if (selected && parseDatasetBoolean(selected.dataset.requireExplanation)) {
+        noteField.required = true;
+        if (label) {
+          // Add a small indicator span instead of replacing the label content
+          let indicator = label.querySelector(".required-indicator");
+          if (!indicator) {
+            indicator = document.createElement("span");
+            indicator.className = "required-indicator";
+            indicator.textContent = " (required)";
+            label.appendChild(indicator);
+          }
+        }
+      } else {
+        noteField.required = false;
+        if (label) {
+          const indicator = label.querySelector(".required-indicator");
+          if (indicator) indicator.remove();
+        }
+      }
+    } catch (err) {
+      console.error("PostFlags.updateNoteRequired error", err);
+    }
+  }
+
+  function showError (msg) {
+    try {
+      if (!noteField) return;
+      if (!errorMessage) {
+        errorMessage = document.createElement("div");
+        errorMessage.className = "flag-note-error error-text";
+        noteField.closest(".flag-notes")?.appendChild(errorMessage);
+      }
+      errorMessage.textContent = msg;
+    } catch (err) {
+      console.error("PostFlags.showError error", err);
+    }
+  }
+
+  function clearError () {
+    if (errorMessage) errorMessage.textContent = "";
+  }
+
+  function isNoteRequiredAndEmpty () {
+    try {
+      if (!noteField) return false;
+      const selected = document.querySelector("input[name=\"post_flag[reason_name]\"]:checked");
+      return selected && parseDatasetBoolean(selected.dataset.requireExplanation) && noteField.value.trim() === "";
+    } catch (err) {
+      console.error("PostFlags.isNoteRequiredAndEmpty error", err);
+      return false;
+    }
+  }
+
+  if (form) {
+    console.debug("PostFlags: attaching submit handler to form", form);
+    form.addEventListener("submit", function (e) {
+      try {
+        updateNoteRequired();
+        if (isNoteRequiredAndEmpty()) {
+          e.preventDefault();
```

Comment:
This code is buggy, because it will prevent the submit event, but Rails UJS will disable the button on submit, to prevent double-submits. This means, we can no longer submit after receiving this error. 
To fix this, we can prevent Rails from doing this 
```ruby
      <%= f.submit "Submit", class: "st-button submit", data: { disable_with: false } %>
```
and do it ourselves:
```js
      } else {
        clearError();
        e.submitter.disabled = true;
      }
```